{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__This portion of the code deals with aligning the 5'-UTR portion of sequences__\n",
    "\n",
    "Sequences that are aligned must have a 5_UTR of atleast 100 nucleotides so that an alignment can be produced. Filtering steps that were included in the previous section to make Synplot2 run are not included here or in downstream analyses\n",
    "\n",
    "__Mammalia Results__\n",
    "\n",
    "\n",
    "The sequences that lack a conserved CUG codon are Camelus_ferus, Vombatus_ursinus, Phascolarctos_cinereus , Monodelphis_domestica. \n",
    "\n",
    "Sarcophilus_harrisii has the CUG codon but the surrounding sequence is a bit odd\n",
    "\n",
    "Unsurprisingly, the length of the alternative frame in the CDS for sequences are the following \n",
    "\n",
    "Camelus_ferus: 197 nucleotides\n",
    "Sarcophilus_harrisii: 86 nucleotides\n",
    "Vombatus_ursinus: 86 nucleotides\n",
    "Phascolarctos_cinereus: 86 nucleotides\n",
    "Monodelphis_domestica: 86 nucleotides\n",
    "\n",
    "This supports the idea that the mammals that lack the CUG initiation codon, also do not have the long ORF either\n",
    "\n",
    "Monodelphis_domestica has a gap where the CUG should be located in mammals so I've artificially just chosen the next nucleotide as where the reference location begins. This can be seen by referring to the KozakMotif function\n",
    "\n",
    "\n",
    "__Sauropsid Results__\n",
    "\n",
    "Most sequences do not appear to have a long ORF in the incorrect reading frame. Unlike mammalia, I may have to manually pick and choose sequences that work best for me to align since an alignment with ALL the kozak sequences isn't yielding anything informative. Perhaps a manual inspection is in order.\n",
    "\n",
    "The peptide generated by a CUG uORF by Gallus gallus has terrible protein alignment to the Homo sapiens protein\n",
    "\n",
    "--> Sequences must have a 3'-UTR of 100 nucleotides or greater\n",
    "--> The extension in the correct ORF must be greater than 300 nucleotides\n",
    "\n",
    "_Gallus_gallus_\n",
    "\n",
    "-->CUG is properly offset from start codon\n",
    "-->Length of ORF from CUG to TF stop codon is 1110 nucleotides\n",
    "\n",
    "Empidonax_traillii\n",
    "\n",
    "-->CUG is not properly offset from start codon to access frame\n",
    "-->Length of start codon to supposed TF stop codon is 398 nucleotides\n",
    "\n",
    "Neopelma_chrysocephalum\n",
    "-->CUG is not properly offset from start codon to access frame\n",
    "-->Length of start codon to supposed TF stop codon is 398 nucleotides\n",
    "\n",
    "_Parus_major_\n",
    "-->CUG is properly offset from start codon\n",
    "-->Length of ORF from CUG to TF stop codon is 777 nucleotides\n",
    "\n",
    "_Zonotrichia_albicollis_\n",
    "-->CUG is properly offset from the start codon\n",
    "-->Length of ORF from CUG to TF stop codon is 813 nucleotides\n",
    "\n",
    "Corvus_brachyrhynchos\n",
    "-->CUG is not properly offset from the start codon\n",
    "-->Length of ORF from start codon to supposed TF stop codon is 536 nucleotides\n",
    "\n",
    "_Corapipo_altera_\n",
    "-->CUG is properly offset from the start codon\n",
    "-->Length of ORF from CUG to TF stop codon is 471 nucleotides\n",
    "\n",
    "_Lepidothrix_coronata_\n",
    "-->CUG is properly offset from the start codon\n",
    "-->Length of ORF from CUG to TF stop codon is 471 nucleotides\n",
    "\n",
    "_Pipra_filicauda_\n",
    "-->CUG is properly offset from the start codon\n",
    "-->Length of ORF from CUG to TF stop codon is 471 nucleotides\n",
    "\n",
    "Ficedula_albicollis\n",
    "-->CUG is not properly offset from the start codon\n",
    "-->Length of ORF from start to supposed stop codon is 353 nucleotides\n",
    "\n",
    "_Anser_cygnoides_domesticus_\n",
    "-->There are two CUGs that are properly offset from the start codon. One is 100 nucleotides away and the other is 178 nucleotides away. I assume that the one that is 100 nucleotides is likely to be the one used (or formerly used). \n",
    "-->Length of ORF from CUG to TF stop codon is 708 nucleotides\n",
    "\n",
    "\n",
    "List of sauropsids that require further analysis: sauropsid_list = ['Gallus_gallus', 'Parus_major', 'Zonotrichia_albicollis', 'Corapipo_altera', 'Lepidothrix_coronata', 'Pipra_filicauda', 'Anser_cygnoides_domesticus']\n",
    "\n",
    "These sauropsids have no alignment with their prospective CUG initiation codon in a MUSCLE alignment of their 5UTRs\n",
    "\n",
    "__Amphibia__\n",
    "\n",
    "There are only three organisms that are analyzed. None have an overlapping extension greater than 300 nucleotides\n",
    "\n",
    "Xenopus_tropicalis:XM_002932235\n",
    "Length into TF ORF 71 nucleotides\n",
    "\n",
    "Xenopus_laevis:XM_018250789\n",
    "Length into TF ORF 62 nucleotides\n",
    "\n",
    "Nanorana_parkeri: XM_018571894\n",
    "Length into TF ORF 38 nucleotides\n",
    "\n",
    "\n",
    "__Teleost fish__\n",
    "\n",
    "Only 3 sequences have an overlapping extension greater than 300 nucleotides\n",
    "\n",
    "Austrofundulus_limnaeus (XM_014005514): 323 nucleotides\n",
    "Contains two CUGs in the proper frame 70 and 82 nucleotides away but there is an intervening stop codon. \n",
    "\n",
    "Sinocyclocheilus_rhinocerous (XM_016528575): 314 nucleotides\n",
    "Contains one CUG in the proper frame but there is an intervening stop codon\n",
    "\n",
    "Clupea_harengus (XM_012832592): 320 nucleotides\n",
    "Contains one CUG in the proper frame without an intervening stop codon\n",
    "-->The surrounding sequence for this CUG is ACAAAA __CTG__ A\n",
    ">Potential_protein\n",
    "MKRNVAVMVTPLSTTTVPCTGAQEVAKFLGRASPGTGAQLSSEKSKPTQHPDDVCKPAQADIPRIGASIHTGRCGAEYQAPAEASALGQRSCASSRRGAEVARNVWQRH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio.Alphabet import IUPAC\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "% matplotlib inline\n",
    "import numpy as np\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import csv\n",
    "from itertools import islice\n",
    "from Bio.Emboss.Applications import TranalignCommandline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = os.getcwd()+os.sep\n",
    "muscle_executable = cwd+\"muscle3.8.31_i86win32.exe\"\n",
    "BLAST_genbank = 'Representative_Species/tblastn_refseqrna_mammalia_queryhomosapiens_1-12-20.gb'\n",
    "BLAST_textoutput = 'Representative_Species/tblastn_refseqrna_mammalia_queryhomosapiens_1-12-20.txt'\n",
    "tranalign_exe = r\"C:\\mEMBOSS\\Tranalign.exe\"\n",
    "tranalignseq_out = cwd + 'POLG_tranalign_output_80.fasta'\n",
    "fiveUTR_100nt = cwd+'trimmed_5UTR.fasta'\n",
    "aligned_fiveUTR_100nt = cwd+'aligned_trimmed_5UTR.fasta'\n",
    "kozak_file = cwd+'kozak_sequence_aligned.fasta'\n",
    "CUGRNAfasta = cwd + 'CUG_RNA.fasta'\n",
    "CUGalignedRNA = cwd + 'CUG_RNA_aligned.clw'\n",
    "CUG_RNA_aligned_fasta = cwd + 'CUG_RNA_aligned.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is the POLG201 sequence for gallus gallus because the NCBI version has the CUG in the incorrect frame\n",
    "custom_POLG201_Gallus_gallus_sequence = Seq('GTCGCCCCGGAGCCCCGCGTTGCACCGCGATCCGACCCGGGCGGCGCGGTGTGGGGGCGGGGGGGCGCGTGGGGACAGCGCGGGGCTGCACGGCGCGGGGAAGGGGAGGTGCGGAGCTTTGGGGGCGCGTGCAGAGCTGTGGGGAGCGGCGGGGCGTGCCGCGTGCCTTGCAGGGGTGCCGCGTGCTTTGCAGGGGTGCCGCGTGCGTGCAGAGGTGTTGCACGCCTTGCAGGGGTGCTGCGTGCTTCGACGCAGTGGTGCGCCCCAGCTGCGCATCCCGGCGCGCCGCACACCTACGAGGTGTCGCTTTTAGCAGCGCCGGTTGAAGCTCCCGCTGCGGACCCCTCCCCTCACCGCCGCTCTCCCCCTGCAGGAGGACCCCCCCCTTATCCGGCAGCACCGCGATGCTCCGCGCGCTCCGCCGAGGCTCAGCGCCGCGCCGCGCCGCCTCCCGGCCGTGCTCCGGGCCCTCCGCGCACCGCCCGCAGCCACGCGGCGACGAGGCCGAGCCGTCGGAGCGGAGCGAGCGCCGCGTGAACCCGCTGCACATCCAGATGTTGTCCCGGAACCTCCACGAGCAGATCTTCCGCGGGGCGCCCGTGCGGCACTCGGAGGCGGCCGTGCGGCGCAGCGTCGAACACCTGCAGCGGCACGGCCTGTGGGGCCGGCACGGCCCGTCGCTGCCCGACGTGAGTTTGCGCCTGCCCCGCATGTACGGCGCCGACATCGACGAGCATTTCCGCCGCCTGGCGCAGAAGCAGAGCCTGCCCTACCTGGAGGCGGCCGAGGAGCTGCTGCGCTGCCGCCTGCCCCCCGCACCACAGAGCTGGGCCCGGCAGCAGGGCTGGACGCGCTACGGCCCCGACGGGCGGCCCGAGGCGGTGGAGTGCCCGCGGGAGCGCGCGTTGGTGCTGGACGTGGAGGTGTGCGTGGCCGCCGGGCAGTGCCCCACTATGGCCGTGGCGGTGTCGCCGCACGCGTGGTACTCGTGGTGCAGCCGGCGCCTGCTGGAGCAGCGCTACTCGTGGGGCCCCCGGCTGGCGCTGCACGACCTCGTGCCTCTGGAGGGGACCGGCAGGCAGCAGGAGGGCGGCGAGAGGGTGGTGGTGGGGCACAACGTGGCTTTCGACCGCGCCTTCATCAGGGAGCAGTACCTCGTGCAGGGCTCCCGGGTGCGCTTCCTGGACACCATGAGCATGCACATGGCCATCTCGGGGCTCACGGGCTTCCAGCGCAGCCTCTGGATGGCCGCCAAGCACGGCAAGAGGAAGGGGCTGCAGCAGGTCAGGCAGCACATGAAGAAGACACGCAGCAAAGCCGAGGGGCCGGCGGTCTCTTCATGGGACTGGGTGCACGTCAGCAGCATCAACAACCTGGCAGATGTGCATGCACTGTACGTGGGAGGGGAACCGCTGCAGAAGGAGGCACGAGAGCTGTTTGTTAAGGGGACCATGGCTGACGTCAGGAATAACTTCCAGGAGCTGATGTCGTACTGTGCCAGCGATGTCCGGGCCACCTATGAGGTGTTCCAGGAGCAGCTGCCGCTCTTCATGGAGAGGTGCCCCCACCCCGTGACGTTTGCTGGGATGTTGGAGATGGGGGTGTCCTACCTGCCGGTCAACAGCAACTGGAGGAGGTACCTGGACGATGCTCAGGGCACCTATGAGGAGCTGCAGAAGGAGATGAAAAAGTCCTTGATGAACCTGGCCAACGATGCCTGCCAGCTGCTGCACGAGGACAGGTACAAGGAGGACCCCTGGCTCTGGGATCTGGAGTGGGACACGCAAGAGTTTAAGCAGAAGAAACCCGCTAAGAGGAAGAAGGATCAGAAAATAAACAGTGAAGCTTCCGAGACGGGCTCTGCTCAGGAGTGGAGGGAAGACCCCGGTCCCCCCAGCGAGGAGGAGGAGCTGAGAGCCCCCGAGAGCAGCACCTGCCTGGAGCGCCTGAAGGAGACGATCACACTGCAGCCCAAGAGGCTGCAGCACCTCCCGGGCCACCCGGGCTGGTACCGCAAGCTCTGCCCGCGCCTGGAGGAGGAGGGCTGGGTGCCGGGGCCCAGCCTCATCAGCCTGCAGATGCGGGTGACCCCGAAACTGATGCGCCTGGCCTGGGATGGCTTCCCTCTGCACTACTCGGAGAAGCACGGCTGGGGCTACCTGGTGCCGGGGCGGCAGGACAACCTGCCTGCAGCCTCTGCGGAGCCAGAGGGGCCTGTCTGCCCACACAGGGCGATCGAGCGGCTGTATCGGCAGCACTGCCTGCAGAGGGGCCAGGAGCAGCCCCCAGAGGAGGCTGGCGTGGAGGATGAGCTGATGGTGCTGGAGGGCAGCAGCATGTGGCAGAAGGTGGAGGAGCTGAGCCAGCTGGAGCTGGACATGGAGCGGCCGGGCAGGGCAGAGCAGAGCCAGATGCAGGATGAGGACGGGCTGCCAGAGCTGGTGGAGGAGAGCAGCCAGCCCTCATTCCACCACGGCAATGGCCCCTACAACGACGTCAACATCCCTGGATGCTGGTTCTTCAAGCTGCCCCACAAGGACGGCAATGAGAACAACGTGGGGAGCCCCTTTGCCAAGGACTTCCTGCCCCGCATGGAGGATGGCACGCTGCGGGCCACCGTGGGCCGCACCCATGGGACCAGAGCCCTGGAGATCAACAAGATGGTGTCCTTCTGGAGGAACGCTCACAAGCGGGTCAGTTCCCAGGTGGTTGTGTGGCTGAAGAAGGGGGAGCTGCCCCGTGCGGTGACCAGGCACCCGGCCTACAGCGAGGAGGAGGACTACGGGGCCATCCTGCCGCAGGTGGTGACTGCGGGTACCATCACCCGTCGGGCCGTGGAGCCCACGTGGCTGACAGCCAGCAATGCCCGGGCTGACCGTGTGGGCAGCGAGCTGAAGGCCATGGTCCAGGTGCCGCCCGGCTACTCTCTGGTGGGTGCAGATGTGGACTCCCAGGAGCTGTGGATAGCGGCGGTCCTGGGCGAGGCTCACTTTGCTGGCATGCACGGGTGCACGGCCTTCGGCTGGATGACCCTGCAAGGGAAGAAGAGCGACGGGACCGACCTGCATAGCAAGACGGCCGCCACGGTGGGCATCAGCCGGGAGCACGCCAAGGTCTTCAACTACGGGCGCATCTACGGGGCTGGGCAGCCCTTTGCCGAGCGGCTGCTGATGCAGTTCAATCACCGGCTGACACAGCAGCAGGCACGTGAGAAGGCACAGCAGATGTATGCAGTCACAAAGGGCATCCGGAGGTTTCATCTCAGCGAGGAGGGCGAGTGGCTGGTGAAGGAACTGGAGCTGGCTGTGGACAAAGCAGAAGATGGTACGGTGTCGGCCCAGGATGTGCAGAAGATCCAGAGAGAAGCCATGAGAAAGTCCCGAAGGAAGAAGAAGTGGGACGTGGTGGCTCACCGAATGTGGGCTGGAGGCACCGAGTCCGAAATGTTCAACAAGCTGGAGAGCATCGCTCTGTCCGCCTCGCCACAGACCCCGGTGCTGGGCTGTCATATCAGCAGGGCTCTGGAGCCTGCAGTGGCCAAAGGGGAGTTTCTAACCAGCAGAGTGAACTGGGTGGTGCAGAGCTCAGCTGTTGACTACCTGCACCTCATGCTGGTCTCCATGAAGTGGCTCTTTGAGGAGTATGACATAAATGGTCGCTTCTGCATCAGCATCCACGACGAGGTGCGCTACCTGGTGCAGGAGCAGGACCGCTACCGGGCAGCACTGGCCCTGCAGATCACCAACCTGCTCACACGGTGCATGTTTGCCTACAAGCTGGGCCTCCAGGATCTGCCGCAGTCCGTGGCTTTCTTCAGCGCTGTGGACATTGACCGGTGCTTAAGGAAGGAGGTGACCATGAACTGTGCGACTCCATCAAATCCAACCGGCATGGAGAAGAAGTACGGCATTCCTCGAGGAGAAGCACTGGATATATATCAGATAATTGAAATAACCAAAGGCTCACTGGAGAAGAAGTGATAACGTGAGAGTGCCAGAAGGTGCAAGTTGTCCAGAGAGCACACGGGAACCTGGCTGTCCTTTCAGAAGCACATACATGGCAGGGACCAATCCTGGTTGCGCCGCTTCCTTCTCGTGGTAAGAAAAAGATGTTCCTGATGAAGATTTTCATAGCAGCACATCTGAATGGGAGAGCTTGCATATTTGAATGGCTGGCAGCCAGCTTTAAGACCTGAGACACCTGACAGAGTCACTGCTTGCACACCCGTGGGGATGAAGAAAGAAGTCTTGAGTATTTGCCAGGAGACAGAATCAAATCAATCATCTGTACGTGCAGTTCTCCAAGACCAAGGTGAGGCTGCCACAGCACAGGTGCTGTAGGAGAAGGAGGTGGCAGCAGTTGCAAGCACACATTCTATTTTTTTCGCCTTCTTTTCTTTTGGGGTTCCTGGTTTTCATCTGGCTGCTCTGCTGTGCCGGACTGGAGAGAAATAGAGAGTTAAGAGTACCAAGTGTGAACGTTTGTGT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function requires a genbank file from a blast result, the text_table of the results of a blast result. An optional\n",
    "#parameter exists called optional_cutoff --> if an individual sorts sequences by something like query cover before downloading\n",
    "#his/her blast result, the user can choose to stop processing at a specific sequence so that only sequences above a certain\n",
    "#query cover are considered in downstream analysis. Alternatively, one could simply download sequences manually that are above\n",
    "#a certain query cover\n",
    "def processHitTable(genbank_file,text_table, optional_cutoff = ''):\n",
    "    Sequence_dict = {}\n",
    "    for file in SeqIO.parse(genbank_file, 'gb'):\n",
    "        for feature in file.features:\n",
    "                if feature.type == 'gene':\n",
    "                    if 'gene' in feature.qualifiers.keys():\n",
    "                        symbol = feature.qualifiers['gene']\n",
    "                    if 'locus_tag' in feature.qualifiers.keys():\n",
    "                        symbol = feature.qualifiers['locus_tag']\n",
    "                if feature.type == 'source':\n",
    "                    organism = feature.qualifiers['organism'][0].replace(\" \", \"_\")\n",
    "                    \n",
    "                    #automatically should use POLG-201 transcript\n",
    "                    if organism == 'Gallus_gallus':\n",
    "                        Sequence_dict['Gallus_gallus'] = {}\n",
    "                        Sequence_dict['Gallus_gallus']['POLG_201'] = {}\n",
    "                        Sequence_dict['Gallus_gallus']['POLG_201']['nam'] = 'Ensembl_transcript_POLG-201'\n",
    "                        Sequence_dict['Gallus_gallus']['POLG_201']['seq'] = (custom_POLG201_Gallus_gallus_sequence)\n",
    "                        Sequence_dict['Gallus_gallus']['POLG_201']['start'] = 404\n",
    "                        Sequence_dict['Gallus_gallus']['POLG_201']['end'] = 3983\n",
    "                        Sequence_dict['Gallus_gallus']['POLG_201']['bit score'] = 1000000\n",
    "                        \n",
    "                if feature.type == 'CDS':\n",
    "                    CDS = [int(a) for a in feature.location]\n",
    "                    start = CDS[0]\n",
    "                    end = CDS[-1]\n",
    "                    accession = file.name\n",
    "                    full_name = file.description\n",
    "                    if organism not in Sequence_dict.keys():\n",
    "                        Sequence_dict[organism] = dict()\n",
    "                    Sequence_dict[organism][accession] = dict()\n",
    "                    Sequence_dict[organism][accession]['nam'] = full_name\n",
    "                    Sequence_dict[organism][accession]['seq'] = file.seq\n",
    "                    Sequence_dict[organism][accession]['start'] = start\n",
    "                    Sequence_dict[organism][accession]['end'] = end + 1\n",
    "    \n",
    "    final_hit_dict = {}\n",
    "    with open(text_table) as f:\n",
    "        reader = csv.DictReader(f, delimiter = \"\\t\")\n",
    "        for initial_row in islice(reader, 4, 5):\n",
    "            header_list = str((initial_row['# tblastn'])).split('# Fields: ')[1].split(', ')\n",
    "        hit_number = 0\n",
    "        for row in islice(reader, 1, None):   \n",
    "            hit_dict = {}\n",
    "            hit_number +=1\n",
    "            query_id = []\n",
    "            query_id.append(str(row['# tblastn']))\n",
    "            result_list = row[None]\n",
    "            combined_results = query_id + result_list\n",
    "            i = 0\n",
    "            for item in header_list:\n",
    "                hit_dict[item] = combined_results[i]\n",
    "                i+=1  \n",
    "            if optional_cutoff != '':\n",
    "                if hit_dict['subject acc.ver'] == optional_cutoff:\n",
    "                    break\n",
    "            key = (hit_dict['subject acc.ver'].split('.'))[0]\n",
    "            \n",
    "            \n",
    "            organism = ''\n",
    "            accession = ''\n",
    "            for item in Sequence_dict:\n",
    "                for item2 in Sequence_dict[item]:\n",
    "                    if item2 == key:\n",
    "                        organism = item\n",
    "                        accession = item2\n",
    "                        Sequence_dict[organism][accession].update(hit_dict)\n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "            #final_hit_dict[key] = hit_dict     \n",
    "    return Sequence_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bestHitPerOrganism(hitTable):\n",
    "    singleHitDict = {}\n",
    "    for organism in hitTable:\n",
    "        bestScore = 0.0\n",
    "        final_accession = ''\n",
    "        transcript_variant = 0\n",
    "        for accession in hitTable[organism]:\n",
    "            current_score = float(hitTable[organism][accession]['bit score'])\n",
    "            if current_score > bestScore:\n",
    "                bestScore = current_score\n",
    "                final_accession = accession\n",
    "        singleHitDict[organism] = {'accession':final_accession, 'bit_score': bestScore,\n",
    "                                   'sequence':hitTable[organism][final_accession]['seq'],\n",
    "                                  'nam':hitTable[organism][final_accession]['nam'],\n",
    "                                  'start':hitTable[organism][final_accession]['start'],\n",
    "                                  'end':hitTable[organism][final_accession]['end']}\n",
    "    return singleHitDict     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "hitTable = processHitTable(BLAST_genbank,BLAST_textoutput)\n",
    "singleTranscriptTable = bestHitPerOrganism(hitTable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extractfiveUTR(singleTranscriptTable, UTR_size = 100):\n",
    "    five_UTR_dict = {}\n",
    "    for item in singleTranscriptTable:\n",
    "        if singleTranscriptTable[item]['start'] > UTR_size:\n",
    "            accession = singleTranscriptTable[item]['accession']\n",
    "            sequence = singleTranscriptTable[item]['sequence']\n",
    "            fiveUTR = sequence[0:singleTranscriptTable[item]['start']]\n",
    "            five_UTR_dict[item] = {'accession':accession,'fiveUTR':fiveUTR}\n",
    "    return five_UTR_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fiveUTRTrim(five_UTR_dict, trim_size = 100):\n",
    "    UTR_trim_dict = {}\n",
    "    for item in five_UTR_dict:\n",
    "        len_UTR = len(five_UTR_dict[item]['fiveUTR'])\n",
    "        start = len_UTR-trim_size\n",
    "        trimmed_UTR = five_UTR_dict[item]['fiveUTR'][start:len_UTR]\n",
    "        UTR_trim_dict[item] = {'accession':five_UTR_dict[item]['accession'],'sequence':trimmed_UTR}\n",
    "    return UTR_trim_dict\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trimmedfiveUTRwrite(UTR_trim_dict):\n",
    "    UTR_file = open(fiveUTR_100nt,'w')\n",
    "    for item in UTR_trim_dict:\n",
    "        UTR_file.write('>'+item+'\\n'+str(UTR_trim_dict[item]['sequence'])+'\\n')\n",
    "    UTR_file.close()\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "def determineAltFrameLength(singleTranscriptTable):\n",
    "    plus1_dict = {}\n",
    "    for item in singleTranscriptTable:\n",
    "        sequence = singleTranscriptTable[item]['sequence']\n",
    "        start = singleTranscriptTable[item]['start']\n",
    "        stop = singleTranscriptTable[item]['end']\n",
    "        accession = singleTranscriptTable[item]['accession']\n",
    "        CDS = sequence[start:stop]\n",
    "        CDS_truncateStart = CDS[3:]\n",
    "        plusOne = 'ATGG' + CDS_truncateStart\n",
    "        plusOneLength = (3*len(plusOne.translate(to_stop=True)))+2\n",
    "        plus1_dict[item] = {'accession':accession,'+1_length_intoCDS':plusOneLength}\n",
    "    return plus1_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kozakMotif(fiveUTRalignment_dict,reference_location):\n",
    "    reference_location -=1\n",
    "    kozak_dict = {}\n",
    "    for item in fiveUTRalignment_dict:\n",
    "        if item == 'Monodelphis_domestica':\n",
    "            reference_location +=1\n",
    "        \n",
    "        \n",
    "        fiveUTR = fiveUTRalignment_dict[item]\n",
    "        CUG = fiveUTR[reference_location:reference_location+3]\n",
    "        motif = CUG\n",
    "        i = 0\n",
    "        nextnt = fiveUTR[reference_location+3+i:reference_location+3+i+1]\n",
    "        while nextnt == '-':\n",
    "            i +=1\n",
    "        motif += nextnt\n",
    "    \n",
    "        k = 0\n",
    "        j = 6\n",
    "    \n",
    "        while j > 0:\n",
    "            previousnt = fiveUTR[reference_location-1+k:reference_location+k]\n",
    "            if previousnt == '-':\n",
    "                k -=1\n",
    "            else:\n",
    "                motif = previousnt + motif\n",
    "                k -=1\n",
    "                j -=1\n",
    "        kozak_dict[item] = motif\n",
    "    return kozak_dict\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "def runMuscle(in_file, out_file, muscle_executable):\n",
    "    muscle_cline = MuscleCommandline(muscle_executable, input=in_file, out=out_file)\n",
    "    muscle_cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeKozakMotif(kozak_motif_dict, kozak_file):\n",
    "    file = open(kozak_file, 'w')\n",
    "    for item in kozak_motif_dict:\n",
    "        file.write('>'+item+'\\n'+kozak_motif_dict[item]+'\\n')\n",
    "    file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readAlignment(in_file):\n",
    "    alignment_dict = {}\n",
    "    for seq_record in (SeqIO.parse(in_file, 'fasta')):\n",
    "        name = seq_record.id\n",
    "        seq = seq_record.seq\n",
    "        alignment_dict[name] = str(seq)\n",
    "    return (alignment_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "five_UTR_dict = extractfiveUTR(singleTranscriptTable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "UTR_trim_dict = fiveUTRTrim(five_UTR_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "trimmedfiveUTRwrite(UTR_trim_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "runMuscle(fiveUTR_100nt, aligned_fiveUTR_100nt, muscle_executable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "fiveUTRalignment_dict = readAlignment(aligned_fiveUTR_100nt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Maste\\Anaconda3\\lib\\site-packages\\Bio\\Seq.py:2576: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "plus1_dict = determineAltFrameLength(singleTranscriptTable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "#custom_number that will depend on the alignment and requires manual insepction\n",
    "mammalian_CUG_location = 86"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "kozak_motif_dict = kozakMotif(fiveUTRalignment_dict,mammalian_CUG_location)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This will output a file with your aligned sequences with the kozak sequence\n",
    "writeKozakMotif(kozak_motif_dict, kozak_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Looking at RNA secondary structure between CUG and AUG start codon__\n",
    "\n",
    "__Mammalia__\n",
    "\n",
    "I'm going to use the same sequences that I used for the protein analysis. If the CUG is in the p-site, then the following 3 nucleotides will be in the A-site. The following 5 nucleotides will be in the entry tunnel of the ribosome. Thus I should only consider the nucleotides 8 away from the CUG intiiation site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "mammal_ignore_list = ['Camelus_ferus', 'Vombatus_ursinus', 'Phascolarctos_cinereus' ,'Monodelphis_domestica']\n",
    "mammal_ignore_list_extended = ['Camelus_ferus', 'Vombatus_ursinus', 'Phascolarctos_cinereus' ,'Monodelphis_domestica','Sarcophilus_harrisii']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCUGRNA(UTR_trim_dict,ignore_list,singleTranscriptTable):\n",
    "    \n",
    "    CUG_RNA_dict = {}\n",
    "    for item in UTR_trim_dict:\n",
    "        if item in ignore_list:\n",
    "            continue\n",
    "        fiveUTRcode = fiveUTRalignment_dict[item][86:]\n",
    "        fiveUTRCDS = ''\n",
    "        for nt in fiveUTRcode:\n",
    "            if nt != '-':\n",
    "                fiveUTRCDS = fiveUTRCDS+ nt\n",
    "        fiveUTRCDS = fiveUTRCDS[10:]\n",
    "        CUG_RNA_dict[item] = Seq(fiveUTRCDS).transcribe()\n",
    "    return CUG_RNA_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeRNA(RNA_dict, outfile):\n",
    "    file = open(outfile,'w')\n",
    "    for item in RNA_dict:\n",
    "        file.write('>'+item+'\\n'+str(RNA_dict[item])+'\\n')\n",
    "    file.close()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generateAlignmentFile(alignment_in, alignment_dict):\n",
    "    alignment_file = open(alignment_in, 'w')\n",
    "    for item in alignment_dict:\n",
    "        alignment_file.write('>'+item+'\\n'+(alignment_dict[item])+'\\n')\n",
    "    alignment_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "CUGRNA_dict = getCUGRNA(UTR_trim_dict,mammal_ignore_list_extended,singleTranscriptTable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "writeRNA(CUGRNA_dict, CUGRNAfasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "runMuscle(CUGRNAfasta, CUGalignedRNA, muscle_executable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "CUG_RNA_alignment_dict = readAlignment(CUGalignedRNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "generateAlignmentFile(CUG_RNA_aligned_fasta, CUG_RNA_alignment_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
