{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Analysis of ribosome profiling__\n",
    "\n",
    "This data was downloaded from Tripz-Viz on 10-1-19 from the following link (https://trips.ucc.ie/homo_sapiens/Gencode_v25/interactive_plot/?files=&ribo_studies=18,20,21,23,24,27,28,29,31,32,33,34,35,38,39,42,43,44,45,56,58,60,62,63,64,67,89,90,99,101,102,103,107,113,122,124,130,134,138,141,144,150,152,153,165,171,172,176,177,178,179,181,183,190,191,192,201,204,212,&tran=ENST00000268124&minread=25&maxread=150&user_dir=fiveprime&ambig=F&cov=F&lg=T&nuc=F&rs=0&crd=F&short=mcr) with the following transcript identifier (ENST00000268124)\n",
    "\n",
    "Replace the file in riboseq_table with whatever file name you have chosen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio.Alphabet import IUPAC\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "% matplotlib inline\n",
    "import numpy as np\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import csv\n",
    "from itertools import islice\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = os.getcwd()+os.sep\n",
    "riboseq_table = cwd + 'Representative_Species\\Counts.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def processRiboSeqcsv(riboseq_file):\n",
    "    riboseq_dict = {}\n",
    "    Position_list = []\n",
    "    Sequence_list = []\n",
    "    Frame1_list = []\n",
    "    Frame2_list = []\n",
    "    Frame3_list = []\n",
    "    RNASeq_list = []\n",
    "    with open(riboseq_file) as f:\n",
    "            reader = csv.DictReader(f, delimiter = \",\")\n",
    "            for row in reader:\n",
    "                Position_list.append(row['ï»¿Position'])\n",
    "                Sequence_list.append(row['Sequence'])\n",
    "                Frame1_list.append(row['Frame 1'])\n",
    "                Frame2_list.append(row['Frame 2'])\n",
    "                Frame3_list.append(row['Frame 3'])\n",
    "                RNASeq_list.append(row['RNA-Seq'])\n",
    "            \n",
    "    Riboseq_dict = {'Position':Position_list,'Sequence':Sequence_list,\n",
    "                   'Frame1':Frame1_list,'Frame2':Frame2_list,'Frame3':Frame3_list,'RNA_Seq':RNASeq_list}\n",
    "    return Riboseq_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def queryRiboSeq(sequence,Riboseq_dict):\n",
    "    full_length_sequence = ''\n",
    "    for nt in Riboseq_dict['Sequence']:\n",
    "        full_length_sequence += nt\n",
    "    if full_length_sequence.count(sequence) == 0:\n",
    "        print('Sequence was not found')\n",
    "        return\n",
    "    if full_length_sequence.count(sequence) !=1:\n",
    "        print('This sequence occurs more than once, try again with a unique sequence')\n",
    "        return\n",
    "    if full_length_sequence.count(sequence) ==1:\n",
    "        beg_index = 1+full_length_sequence.find(sequence)\n",
    "        end_index = beg_index + len(sequence)\n",
    "        \n",
    "    print(beg_index)\n",
    "    print(end_index)\n",
    "    subsequence_dict = {}\n",
    "    index_counter = beg_index\n",
    "    for nt in sequence:\n",
    "        if index_counter == end_index:\n",
    "            continue\n",
    "        \n",
    "        Frame1_count = Riboseq_dict['Frame1'][index_counter-1]\n",
    "        Frame2_count = Riboseq_dict['Frame2'][index_counter-1]\n",
    "        Frame3_count = Riboseq_dict['Frame3'][index_counter-1]\n",
    "        RNASeq_count = Riboseq_dict['RNA_Seq'][index_counter-1]\n",
    "        \n",
    "        subsequence_dict[index_counter] = {}\n",
    "        subsequence_dict[index_counter]['Nucleotide'] = {nt}\n",
    "        subsequence_dict[index_counter]['Frame1'] = {int(Frame1_count)}\n",
    "        subsequence_dict[index_counter]['Frame2'] = {int(Frame2_count)}\n",
    "        subsequence_dict[index_counter]['Frame3'] = {int(Frame3_count)}\n",
    "        subsequence_dict[index_counter]['RNASeq'] = {int(RNASeq_count)}\n",
    "        index_counter +=1\n",
    "    return subsequence_dict\n",
    "        \n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def printRiboSeqAnalysis(sequence_dict):\n",
    "    full_subsequence = ''\n",
    "    x_total_fr1_counts = float(0)\n",
    "    x_total_fr2_counts = float(0)\n",
    "    x_total_fr3_counts = float(0)\n",
    "    total_fr1_counts = float(0)\n",
    "    total_fr2_counts = float(0)\n",
    "    total_fr3_counts = float(0)\n",
    "    total_all_counts = float(0)\n",
    "    for index in sequence_dict:\n",
    "        for item in sequence_dict[index]['Nucleotide']:\n",
    "            nt = item\n",
    "        full_subsequence += nt\n",
    "        for item in sequence_dict[index]['Frame1']:\n",
    "            fr1_count = float(item)\n",
    "        for item in sequence_dict[index]['Frame2']:\n",
    "            fr2_count = float(item)\n",
    "        for item in sequence_dict[index]['Frame3']:\n",
    "            fr3_count = float(item)\n",
    "        for item in sequence_dict[index]['RNASeq']:\n",
    "            RNASeq_count = float(item)\n",
    "            if RNASeq_count == 0:\n",
    "                RNASeq_count = 1\n",
    "        \n",
    "        \n",
    "        x_total_fr1_counts += (fr1_count/RNASeq_count)\n",
    "        x_total_fr2_counts += (fr2_count/RNASeq_count)\n",
    "        x_total_fr3_counts += (fr3_count/RNASeq_count)\n",
    "    total_fr1_counts = x_total_fr1_counts/float(len(full_subsequence))\n",
    "    total_fr2_counts = x_total_fr2_counts/float(len(full_subsequence))\n",
    "    total_fr3_counts = x_total_fr3_counts/float(len(full_subsequence))\n",
    "    total_all_counts = total_fr1_counts+total_fr2_counts+total_fr3_counts\n",
    "    #counts_per_nt = total_all_counts/float(len(full_subsequence))\n",
    "    print('For the sequence %s, Frame 1 has %f reads, Frame 2 has %f reads, Frame 3 has %f reads, and there are a total of %f reads. '%(full_subsequence,total_fr1_counts,total_fr2_counts,total_fr3_counts,total_all_counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The sequences here are for Homo sapiens. If you choose to do a different organism, obviously this sequences need to be changed\n",
    "\n",
    "uORF_sequence = 'ATGGTCAAACCCATTTCACTGACAGGAGAGCAGAGACAGGACGTGTCTCTCTCCACGTCTTCCAGCCAGTAAA'\n",
    "CTG_sequence = 'CTGGAGCCCAAAGCCAGGTGTTCTGACTCCCAGCGTGGGGGTCCCTGCACCAACCATGAGCCGCCTGCTCTGGAGGAAGGTGGCCGGCGCCACCGTCGGGCCAGGGCCGGTTCCAGCTCCGGGGCGCTGGGTCTCCAGCTCCGTCCCCGCGTCCGACCCCAGCGACGGGCAGCGGCGGCGGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAACAGCAGCCTCAGCAGCCGCAAGTGCTATCCTCGGAGGGCGGGCAGCTGCGGCACAACCCATTGGACATCCAGATGCTCTCGAGAGGGCTGCACGAGCAAATCTTCGGGCAAGGAGGGGAGATGCCTGGCGAGGCCGCGGTGCGCCGCAGCGTCGAGCACCTGCAGAAGCACGGGCTCTGGGGGCAGCCAGCCGTGCCCTTGCCCGACGTGGAGCTGCGCCTGCCGCCCCTCTACGGGGACAACCTGGACCAGCACTTCCGCCTCCTGGCCCAGAAGCAGAGCCTGCCCTACCTGGAGGCGGCCAACTTGCTGTTGCAGGCCCAGCTGCCCCCGAAGCCCCCGGCTTGGGCCTGGGCGGAGGGCTGGACCCGGTACGGCCCCGAGGGGGAGGCCGTACCCGTGGCCATCCCCGAGGAGCGGGCCCTGGTGTTCGACGTGGAGGTCTGCTTGGCAGAGGGAACTTGCCCCACATTGGCGGTGGCCATATCCCCCTCGGCCTGGTATTCCTGGTGCAGCCAGCGGCTGGTGGAAGAGCGTTACTCTTGGACCAGCCAGCTGTCGCCGGCTGA'\n",
    "Main_ORF= 'ATGAGCCGCCTGCTCTGGAGGAAGGTGGCCGGCGCCACCGTCGGGCCAGGGCCGGTTCCAGCTCCGGGGCGCTGGGTCTCCAGCTCCGTCCCCGCGTCCGACCCCAGCGACGGGCAGCGGCGGCGGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAACAGCAGCCTCAGCAGCCGCAAGTGCTATCCTCGGAGGGCGGGCAGCTGCGGCACAACCCATTGGACATCCAGATGCTCTCGAGAGGGCTGCACGAGCAAATCTTCGGGCAAGGAGGGGAGATGCCTGGCGAGGCCGCGGTGCGCCGCAGCGTCGAGCACCTGCAGAAGCACGGGCTCTGGGGGCAGCCAGCCGTGCCCTTGCCCGACGTGGAGCTGCGCCTGCCGCCCCTCTACGGGGACAACCTGGACCAGCACTTCCGCCTCCTGGCCCAGAAGCAGAGCCTGCCCTACCTGGAGGCGGCCAACTTGCTGTTGCAGGCCCAGCTGCCCCCGAAGCCCCCGGCTTGGGCCTGGGCGGAGGGCTGGACCCGGTACGGCCCCGAGGGGGAGGCCGTACCCGTGGCCATCCCCGAGGAGCGGGCCCTGGTGTTCGACGTGGAGGTCTGCTTGGCAGAGGGAACTTGCCCCACATTGGCGGTGGCCATATCCCCCTCGGCCTGGTATTCCTGGTGCAGCCAGCGGCTGGTGGAAGAGCGTTACTCTTGGACCAGCCAGCTGTCGCCGGCTGACCTCATCCCCCTGGAGGTCCCTACTGGTGCCAGCAGCCCCACCCAGAGAGACTGGCAGGAGCAGTTAGTGGTGGGGCACAATGTTTCCTTTGACCGAGCTCATATCAGGGAGCAGTACCTGATCCAGGGTTCCCGCATGCGTTTCCTGGACACCATGAGCATGCACATGGCCATCTCAGGGCTAAGCAGCTTCCAGCGCAGTCTGTGGATAGCAGCCAAGCAGGGCAAACACAAGGTCCAGCCCCCCACAAAGCAAGGCCAGAAGTCCCAGAGGAAAGCCAGAAGAGGCCCAGCGATCTCATCCTGGGACTGGCTGGACATCAGCAGTGTCAACAGTCTGGCAGAGGTGCACAGACTTTATGTAGGGGGGCCTCCCTTAGAGAAGGAGCCTCGAGAACTGTTTGTGAAGGGCACCATGAAGGACATTCGTGAGAACTTCCAGGACCTGATGCAGTACTGTGCCCAGGACGTGTGGGCCACCCATGAGGTTTTCCAGCAGCAGCTACCGCTCTTCTTGGAGAGGTGTCCCCACCCAGTGACTCTGGCCGGCATGCTGGAGATGGGTGTCTCCTACCTGCCTGTCAACCAGAACTGGGAGCGTTACCTGGCAGAGGCACAGGGCACTTATGAGGAGCTCCAGCGGGAGATGAAGAAGTCGTTGATGGATCTGGCCAATGATGCCTGCCAGCTGCTCTCAGGAGAGAGGTACAAAGAAGACCCCTGGCTCTGGGACCTGGAGTGGGACCTGCAAGAATTTAAGCAGAAGAAAGCTAAGAAGGTGAAGAAGGAACCAGCCACAGCCAGCAAGTTGCCCATCGAGGGGGCTGGGGCCCCTGGTGATCCCATGGATCAGGAAGACCTCGGCCCCTGCAGTGAGGAGGAGGAGTTTCAACAAGATGTCATGGCCCGCGCCTGCTTGCAGAAGCTGAAGGGGACCACAGAGCTCCTGCCCAAGCGGCCCCAGCACCTTCCTGGACACCCTGGATGGTACCGGAAGCTCTGCCCCCGGCTAGACGACCCTGCATGGACCCCGGGCCCCAGCCTCCTCAGCCTGCAGATGCGGGTCACACCTAAACTCATGGCACTTACCTGGGATGGCTTCCCTCTGCACTACTCAGAGCGTCATGGCTGGGGCTACTTGGTGCCTGGGCGGCGGGACAACCTGGCCAAGCTGCCGACAGGTACCACCCTGGAGTCAGCTGGGGTGGTCTGCCCCTACAGAGCCATCGAGTCCCTGTACAGGAAGCACTGTCTCGAACAGGGGAAGCAGCAGCTGATGCCCCAGGAGGCCGGCCTGGCGGAGGAGTTCCTGCTCACTGACAATAGTGCCATATGGCAAACGGTAGAAGAACTGGATTACTTAGAAGTGGAGGCTGAGGCCAAGATGGAGAACTTGCGAGCTGCAGTGCCAGGTCAACCCCTAGCTCTGACTGCCCGTGGTGGCCCCAAGGACACCCAGCCCAGCTATCACCATGGCAATGGACCTTACAACGACGTGGACATCCCTGGCTGCTGGTTTTTCAAGCTGCCTCACAAGGATGGTAATAGCTGTAATGTGGGAAGCCCCTTTGCCAAGGACTTCCTGCCCAAGATGGAGGATGGCACCCTGCAGGCTGGCCCAGGAGGTGCCAGTGGGCCCCGTGCTCTGGAAATCAACAAAATGATTTCTTTCTGGAGGAACGCCCATAAACGTATCAGCTCCCAGATGGTGGTGTGGCTGCCCAGGTCAGCTCTGCCCCGTGCTGTGATCAGGCACCCCGACTATGATGAGGAAGGCCTCTATGGGGCCATCCTGCCCCAAGTGGTGACTGCCGGCACCATCACTCGCCGGGCTGTGGAGCCCACATGGCTCACCGCCAGCAATGCCCGGCCTGACCGAGTAGGCAGTGAGTTGAAAGCCATGGTGCAGGCCCCACCTGGCTACACCCTTGTGGGTGCTGATGTGGACTCCCAAGAGCTGTGGATTGCAGCTGTGCTTGGAGACGCCCACTTTGCCGGCATGCATGGCTGCACAGCCTTTGGGTGGATGACACTGCAGGGCAGGAAGAGCAGGGGCACTGATCTACACAGTAAGACAGCCACTACTGTGGGCATCAGCCGTGAGCATGCCAAAATCTTCAACTACGGCCGCATCTATGGTGCTGGGCAGCCCTTTGCTGAGCGCTTACTAATGCAGTTTAACCACCGGCTCACACAGCAGGAGGCAGCTGAGAAGGCCCAGCAGATGTACGCTGCCACCAAGGGCCTCCGCTGGTATCGGCTGTCGGATGAGGGCGAGTGGCTGGTGAGGGAGTTGAACCTCCCAGTGGACAGGACTGAGGGTGGCTGGATTTCCCTGCAGGATCTGCGCAAGGTCCAGAGAGAAACTGCAAGGAAGTCACAGTGGAAGAAGTGGGAGGTGGTTGCTGAACGGGCATGGAAGGGGGGCACAGAGTCAGAAATGTTCAATAAGCTTGAGAGCATTGCTACGTCTGACATACCACGTACCCCGGTGCTGGGCTGCTGCATCAGCCGAGCCCTGGAGCCCTCGGCTGTCCAGGAAGAGTTTATGACCAGCCGTGTGAATTGGGTGGTACAGAGCTCTGCTGTTGACTACTTACACCTCATGCTTGTGGCCATGAAGTGGCTGTTTGAAGAGTTTGCCATAGATGGGCGCTTCTGCATCAGCATCCATGACGAGGTTCGCTACCTGGTGCGGGAGGAGGACCGCTACCGCGCTGCCCTGGCCTTGCAGATCACCAACCTCTTGACCAGGTGCATGTTTGCCTACAAGCTGGGTCTGAATGACTTGCCCCAGTCAGTCGCCTTTTTCAGTGCAGTCGATATTGACCGGTGCCTCAGGAAGGAAGTGACCATGGATTGTAAAACCCCTTCCAACCCAACTGGGATGGAAAGGAGATACGGGATTCCCCAGGGTGAAGCGCTGGATATTTACCAGATAATTGAACTCACCAAAGGCTCCTTGGAAAAACGAAGCCAGCCTGGACCATAG'\n",
    "Overlap_ORF = 'ATGAGCCGCCTGCTCTGGAGGAAGGTGGCCGGCGCCACCGTCGGGCCAGGGCCGGTTCCAGCTCCGGGGCGCTGGGTCTCCAGCTCCGTCCCCGCGTCCGACCCCAGCGACGGGCAGCGGCGGCGGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAACAGCAGCCTCAGCAGCCGCAAGTGCTATCCTCGGAGGGCGGGCAGCTGCGGCACAACCCATTGGACATCCAGATGCTCTCGAGAGGGCTGCACGAGCAAATCTTCGGGCAAGGAGGGGAGATGCCTGGCGAGGCCGCGGTGCGCCGCAGCGTCGAGCACCTGCAGAAGCACGGGCTCTGGGGGCAGCCAGCCGTGCCCTTGCCCGACGTGGAGCTGCGCCTGCCGCCCCTCTACGGGGACAACCTGGACCAGCACTTCCGCCTCCTGGCCCAGAAGCAGAGCCTGCCCTACCTGGAGGCGGCCAACTTGCTGTTGCAGGCCCAGCTGCCCCCGAAGCCCCCGGCTTGGGCCTGGGCGGAGGGCTGGACCCGGTACGGCCCCGAGGGGGAGGCCGTACCCGTGGCCATCCCCGAGGAGCGGGCCCTGGTGTTCGACGTGGAGGTCTGCTTGGCAGAGGGAACTTGCCCCACATTGGCGGTGGCCATATCCCCCTCGGCCTGGTATTCCTGGTGCAGCCAGCGGCTGGTGGAAGAGCGTTACTCTTGGACCAGCCAGCTGTCGCCGGCTGA'\n",
    "Nonoverlap_ORF = 'CCTCATCCCCCTGGAGGTCCCTACTGGTGCCAGCAGCCCCACCCAGAGAGACTGGCAGGAGCAGTTAGTGGTGGGGCACAATGTTTCCTTTGACCGAGCTCATATCAGGGAGCAGTACCTGATCCAGGGTTCCCGCATGCGTTTCCTGGACACCATGAGCATGCACATGGCCATCTCAGGGCTAAGCAGCTTCCAGCGCAGTCTGTGGATAGCAGCCAAGCAGGGCAAACACAAGGTCCAGCCCCCCACAAAGCAAGGCCAGAAGTCCCAGAGGAAAGCCAGAAGAGGCCCAGCGATCTCATCCTGGGACTGGCTGGACATCAGCAGTGTCAACAGTCTGGCAGAGGTGCACAGACTTTATGTAGGGGGGCCTCCCTTAGAGAAGGAGCCTCGAGAACTGTTTGTGAAGGGCACCATGAAGGACATTCGTGAGAACTTCCAGGACCTGATGCAGTACTGTGCCCAGGACGTGTGGGCCACCCATGAGGTTTTCCAGCAGCAGCTACCGCTCTTCTTGGAGAGGTGTCCCCACCCAGTGACTCTGGCCGGCATGCTGGAGATGGGTGTCTCCTACCTGCCTGTCAACCAGAACTGGGAGCGTTACCTGGCAGAGGCACAGGGCACTTATGAGGAGCTCCAGCGGGAGATGAAGAAGTCGTTGATGGATCTGGCCAATGATGCCTGCCAGCTGCTCTCAGGAGAGAGGTACAAAGAAGACCCCTGGCTCTGGGACCTGGAGTGGGACCTGCAAGAATTTAAGCAGAAGAAAGCTAAGAAGGTGAAGAAGGAACCAGCCACAGCCAGCAAGTTGCCCATCGAGGGGGCTGGGGCCCCTGGTGATCCCATGGATCAGGAAGACCTCGGCCCCTGCAGTGAGGAGGAGGAGTTTCAACAAGATGTCATGGCCCGCGCCTGCTTGCAGAAGCTGAAGGGGACCACAGAGCTCCTGCCCAAGCGGCCCCAGCACCTTCCTGGACACCCTGGATGGTACCGGAAGCTCTGCCCCCGGCTAGACGACCCTGCATGGACCCCGGGCCCCAGCCTCCTCAGCCTGCAGATGCGGGTCACACCTAAACTCATGGCACTTACCTGGGATGGCTTCCCTCTGCACTACTCAGAGCGTCATGGCTGGGGCTACTTGGTGCCTGGGCGGCGGGACAACCTGGCCAAGCTGCCGACAGGTACCACCCTGGAGTCAGCTGGGGTGGTCTGCCCCTACAGAGCCATCGAGTCCCTGTACAGGAAGCACTGTCTCGAACAGGGGAAGCAGCAGCTGATGCCCCAGGAGGCCGGCCTGGCGGAGGAGTTCCTGCTCACTGACAATAGTGCCATATGGCAAACGGTAGAAGAACTGGATTACTTAGAAGTGGAGGCTGAGGCCAAGATGGAGAACTTGCGAGCTGCAGTGCCAGGTCAACCCCTAGCTCTGACTGCCCGTGGTGGCCCCAAGGACACCCAGCCCAGCTATCACCATGGCAATGGACCTTACAACGACGTGGACATCCCTGGCTGCTGGTTTTTCAAGCTGCCTCACAAGGATGGTAATAGCTGTAATGTGGGAAGCCCCTTTGCCAAGGACTTCCTGCCCAAGATGGAGGATGGCACCCTGCAGGCTGGCCCAGGAGGTGCCAGTGGGCCCCGTGCTCTGGAAATCAACAAAATGATTTCTTTCTGGAGGAACGCCCATAAACGTATCAGCTCCCAGATGGTGGTGTGGCTGCCCAGGTCAGCTCTGCCCCGTGCTGTGATCAGGCACCCCGACTATGATGAGGAAGGCCTCTATGGGGCCATCCTGCCCCAAGTGGTGACTGCCGGCACCATCACTCGCCGGGCTGTGGAGCCCACATGGCTCACCGCCAGCAATGCCCGGCCTGACCGAGTAGGCAGTGAGTTGAAAGCCATGGTGCAGGCCCCACCTGGCTACACCCTTGTGGGTGCTGATGTGGACTCCCAAGAGCTGTGGATTGCAGCTGTGCTTGGAGACGCCCACTTTGCCGGCATGCATGGCTGCACAGCCTTTGGGTGGATGACACTGCAGGGCAGGAAGAGCAGGGGCACTGATCTACACAGTAAGACAGCCACTACTGTGGGCATCAGCCGTGAGCATGCCAAAATCTTCAACTACGGCCGCATCTATGGTGCTGGGCAGCCCTTTGCTGAGCGCTTACTAATGCAGTTTAACCACCGGCTCACACAGCAGGAGGCAGCTGAGAAGGCCCAGCAGATGTACGCTGCCACCAAGGGCCTCCGCTGGTATCGGCTGTCGGATGAGGGCGAGTGGCTGGTGAGGGAGTTGAACCTCCCAGTGGACAGGACTGAGGGTGGCTGGATTTCCCTGCAGGATCTGCGCAAGGTCCAGAGAGAAACTGCAAGGAAGTCACAGTGGAAGAAGTGGGAGGTGGTTGCTGAACGGGCATGGAAGGGGGGCACAGAGTCAGAAATGTTCAATAAGCTTGAGAGCATTGCTACGTCTGACATACCACGTACCCCGGTGCTGGGCTGCTGCATCAGCCGAGCCCTGGAGCCCTCGGCTGTCCAGGAAGAGTTTATGACCAGCCGTGTGAATTGGGTGGTACAGAGCTCTGCTGTTGACTACTTACACCTCATGCTTGTGGCCATGAAGTGGCTGTTTGAAGAGTTTGCCATAGATGGGCGCTTCTGCATCAGCATCCATGACGAGGTTCGCTACCTGGTGCGGGAGGAGGACCGCTACCGCGCTGCCCTGGCCTTGCAGATCACCAACCTCTTGACCAGGTGCATGTTTGCCTACAAGCTGGGTCTGAATGACTTGCCCCAGTCAGTCGCCTTTTTCAGTGCAGTCGATATTGACCGGTGCCTCAGGAAGGAAGTGACCATGGATTGTAAAACCCCTTCCAACCCAACTGGGATGGAAAGGAGATACGGGATTCCCCAGGGTGAAGCGCTGGATATTTACCAGATAATTGAACTCACCAAAGGCTCCTTGGAAAAACGAAGCCAGCCTGGACCATAG'\n",
    "ORFY_only_nonoverlap = 'CTGGAGCCCAAAGCCAGGTGTTCTGACTCCCAGCGTGGGGGTCCCTGCACCAACC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "Riboseq_dict = processRiboSeqcsv(riboseq_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "280\n",
      "335\n"
     ]
    }
   ],
   "source": [
    "subsequence_dict = queryRiboSeq(ORFY_only_nonoverlap,Riboseq_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the sequence CTGGAGCCCAAAGCCAGGTGTTCTGACTCCCAGCGTGGGGGTCCCTGCACCAACC, Frame 1 has 41.109091 reads, Frame 2 has 16.545455 reads, Frame 3 has 13.781818 reads, and there are a total of 71.436364 reads. \n"
     ]
    }
   ],
   "source": [
    "printRiboSeqAnalysis(subsequence_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
